case-lib/lib.sh

```diff
@@ -1187,3 +1187,42 @@ perf_analyze()
     fi
 }
 
+# Restore the machine state from a file.
+# Should be triggered at the end of
+# every test touching ALSA.
+# Couple with save_machine_state.
+# param1: file name
+restore_alsa_state()
+{
+    dlogi "restore_alsa_state called with ${1}"
+    trap - EXIT
+    dlogi "restore_alsa_state trap removed"
+    trap -p EXIT
+    dlogi "^ restore_alsa_state EXIT trap list"
+    ls /var/tmp | grep ^"$1"
+    dlogi "^ restore_alsa_state, file list"
+    if [ -f "$1" ]; then
+        dlogi "restore_alsa_state found file ${1}"
+        alsactl restore --file /var/tmp/"$1" --pedantic --no-ucm --no-init-fallback
+        dlogi "restore_alsa_state alsactl restore called with ${1}"
+        rm /var/tmp/"$1"
+        dlogi "restore_alsa_state removing file ${1}"
+    fi
+    dlogi "Exiting restore_alsa_state called with ${1}"
+}
+
+# Save the machine state to a file.
+# Should be used at the start of
+# every test touching ALSA.
+# Coupled with restore_machine_state on an exit signal.
+# param1: file name
+save_alsa_state()
+{
+    dlogi "save_alsa_state called with ${1}"
+    trap "restore_alsa_state '${1}'" EXIT
```

Comment:
There is already an EXIT handler: `func_exit_handler()`. You can't have both; how did you test this?

You can likely have `func_exit_handler()` invoke `restore_alsa_state` when the conditions are right.